### DISCUSSION TITLE
What is the difference between Microsoft.Ui.Dispatching.DispatcherQueue and Windows.System.DispatcherQueue?

### DISCUSSION THREAD

**Discussion by brabebhin** [microsoft/microsoft-ui-xaml#7728] [Q&A]
Hi,

I am interested to understand the difference between Microsoft.Ui.Dispatching.DispatcherQueue and Windows.System.DispatcherQueue.

Why can't winui 3 simply use the system dispatcherqueue? 

Thanks.

**Comment by kmgallahan** [comment]
Windows.* = UWP / WinUI 2

Microsoft.* = Win32 / WinUI 3

**Reply by brabebhin** [reply]
Sorry, not what I'm looking for. I need to understand why both exist.

**Reply by kmgallahan** [reply]
WinUI 2 uses the old UWP XAML code that ships with Windows.

WinUI 3 uses new code that ships separately from windows via the Windows App SDK.

They are powered by different code, and said code was setup to use different namespaces. The code behind each DispatcherQueue isn't exactly the same.

**Reply by brabebhin** [reply]
Yes, i can read the adverts myself, thanks.
I need to understand the technical reasons. "it is new code" does not qualify as a reason. I have serious doubts winui is written from scratch. Particularly since the binary interface is the same.

**Reply by DarranRowe** [reply]
Ultimately, there probably isn't very much difference.
The CreateDispatcherQueueController function states:
"If you create a dispatcher queue on the current thread, ensure that there is a message pump running on the current thread so that the dispatcher queue can use it to dispatch tasks."
The Microsoft.UI.Dispatching.DispatcherQueueController.CreateOnCurrentThread static member states:
"Creates a DispatcherQueue on the calling thread which will interop with a USER32 message loop."
Both of these refer to using GetMessage/PeekMessage to get a MSG structure from the thread's message queue. So, they both most likely use a message only window to accomplish this.
The difference then is most likely in how full featured they are. As an example, if you are using Desktop Acrylic or Mica, the documentation tells you to create a Windows.System.DispatcherQueue. It is likely that the Microsoft.UI.Dispatching namespace contains a stripped down version of the dispatcher queue that is enough to get WinUI 3 working on any version of Windows given the whole decoupling it promise, where the system dispatcher queue is fuller featured but is tied to the O/S version.
This is speculation though.

**Reply by brabebhin** [reply]
Thanks for the detailed answer. Cheers!
